### ISSUE TITLE
Support for substate in the state machine visualisers

### ISSUE THREAD

**Issue by JackUkleja** [dotnet-state-machine/stateless#182]
This might be a limitation of the various visualisation languages (e.g. .dot) but it would be great if stateless could render out representations of state machines that make use of SubstateOf()

**Comment by nblumhardt** [comment]
@JackUkleja I think this is done in the 4.0 preview version. Let me know how it goes for you!

**Comment by JackUkleja** [comment]
I ran the latest 4.x preview last Friday and exported my state machine (defined using `SubstateOf`) with the  `UmlDotGraph.Format(sm.GetInfo())` but the graph shows the parent state off to the side and not as a parent of the expected substates. I could be doing something wrong though?

**Comment by HenningNT** [comment]
I remember that someone made a htmlFormatter, maybe it wasn't included in the PR?

**Comment by nblumhardt** [comment]
@betty-crokker ported @sle118's implementation in #169, I believe.

Chris, are we on the right track here? Cheers!

**Comment by betty-crokker** [comment]
@JackUkleja can you update your state machine definition (or a subset thereof)?  As far as I know the code should be working for substates.  What program did you use to visualize your DOT graph?

**Comment by betty-crokker** [comment]
So what do we do, if we never hear back from @JackUkleja ?  Do we continue with the release, and fix this bug later if we figure out what it is?

**Comment by HenningNT** [comment]
Is it a bug, or a missing feature? If it's a missing feature we can release as-is, and add at a later time.

**Comment by betty-crokker** [comment]
The ability to generate DOT graphs of states with substates is definitely in the latest library, and unit tests to verify.  So either:
1) There's a bug in the existing code
2) @JackUkleja has substates in some format/style/layout that the existing code doesn't handle (so deficiency rather than bug)
3) The program @JackUkleja is using to visualize the DOT graphs doesn't handle the way Stateless is currently generating them

**Comment by HenningNT** [comment]
@JackUkleja , could you show us the state machine, and the code you used to generate the dot file?

**Comment by gerardog** [comment]
Guys, this is actually a bug in UmlDotGraphStyle class.

Look at BugTrackerExample on dev branch, this is the current DOT graph.
```
digraph {
compound=true;
node [shape=Mrecord]
rankdir="LR"

subgraph Open
        {
        label = "Open"
Assigned [label="Assigned|exit / Function"];
}
Deferred [label="Deferred|entry / Function"];
Closed [label="Closed"];

Open -> Assigned [style="solid", label="Assign / Function"];
Assigned -> Assigned [style="solid", label="Assign"];
Assigned -> Closed [style="solid", label="Close"];
Assigned -> Deferred [style="solid", label="Defer"];
Deferred -> Assigned [style="solid", label="Assign / Function"];
}
```

It doesnt render the hierarchy, because the subgraph must be named with the "cluster" prefix (lowercase).

This is easy to fix on UmlDotGraphStyle.cs
```
-                 + $"subgraph {stateInfo.NodeName}" + "\n"
+                 + $"subgraph cluster{stateInfo.NodeName}" + "\n"
```

**Comment by betty-crokker** [comment]
Good catch!  Thanks!

**Comment by nblumhardt** [comment]
Closed in #188 - great stuff :-)
